src/main/ruby/truffleruby/core/truffle/io_operations.rb

```diff
@@ -543,10 +543,14 @@ def self.normalize_options(mode, perm, options, default_mode = nil)
             external, internal = encoding.split(':', 2)
           end
         end
+
+        if options[:path]
+          path = options[:path]
+        end
```

Comment:
CRuby converts `path` to String with `#to_str` (and raises error if cannot) and freezes result (but it seems freezing cannot be observer outside of an IO object and it's kind of optimization):

```c
    if (!NIL_P(opt)) {
        if (rb_hash_aref(opt, sym_autoclose) == Qfalse) {
            fmode |= FMODE_PREP;
        }

        path = rb_hash_aref(opt, RB_ID2SYM(idPath));
        if (!NIL_P(path)) {
            StringValue(path);
            path = rb_str_new_frozen(path);
        }
    }
```

Converting to String could be done with helpers like `Truffle::Type.rb_convert_type`

Reply:
Also `if options[:path]` isn't precise enough - it treats `false` similar to `nil` and keeps the `path` variable unassigned. But CRuby will check `nil` value (with `if (!NIL_P(path)) {`) and try to convert `false` to String anyway.

`Primitive.nil?` could be used instead of `#nil?` Ruby method.

Reply:
Ah, good catch. Thanks! 

(Although one wonders who on earth would prefer "false" to be converted to a file path. But API promises are API promises and whatever MRI does is the spec I guess.)